### ISSUE TITLE
External resources are proxies, hence the ontology model cannot be acquired during the mode processing

### ISSUE THREAD

**Issue by kittaakos** [crapo/sadlos2#201]
It cases unresolved reference problem in downstream resources.

**Comment by crapo** [comment]
Pushed changes from this morning's tcon to branch GH-201. Attaching project. Note that the two external OWL file URLs in the .url file will not download for you as they are behind a firewall. Attaching P2 project.
[P2.zip](https://github.com/crapo/sadlos2/files/1307214/P2.zip)

**Comment by crapo** [comment]
One other issue: EmfExternalResource has added dependency on ResourcesPlugin to find the workspace root and convert a platform URI to an absolute File URI. Consequently the tests in ExternalEmfResourceTest mostly fail. Is there a way better way to convert? Or should the tests be moved to ui.tests?

**Comment by kittaakos** [comment]
If it is feasible, we should try to leave the tests in the core.

**Comment by kittaakos** [comment]
@crapo, as I see, all kind of `*.owl` resources get loaded by the external resource factory which I assume, wrong.

Could you please specify what should be the filtering rule? Thank you!

These are a subset of resources which are loaded by the factory:
```
platform:/resource/P2/externalmodels/sadl.sourceforge.net/owl/apvf.owl
platform:/resource/P2/externalmodels/sadl.sourceforge.net/owl/apvf.owl
platform:/resource/P2/OwlModels/extension.sadl.metrics.owl
platform:/resource/P2/OwlModels/SadlImplicitModel.owl
platform:/resource/P2/OwlModels/base.sadl.metrics.owl
...
```

**Comment by kittaakos** [comment]
Could you please confirm, that loading the ontology model instance with the EMF (platform resource) URI as the `base` is correct in the [code](https://github.com/crapo/sadlos2/blob/fc187bddd7035cd3ebaa0f598b0c9e08ed621095/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl/src/com/ge/research/sadl/external/ExternalEmfResource.xtend#L268).

The `base` was unclear to me from the API [documentation](https://jena.apache.org/documentation/javadoc/jena/org/apache/jena/rdf/model/Model.html#read-java.lang.String-java.lang.String-java.lang.String-) of Jena. Thank you!

**Comment by crapo** [comment]
Assuming that one of the first subset of resources is meant to be "...aulo.owl", then the list is correct.

**Comment by kittaakos** [comment]
I do not understand why the implicit model or any generated metric OWL resources have to be handled as external resources? Could you please explain this?

**Comment by crapo** [comment]
I missed the SadlImplicitModel.owl. It should NOT be considered an external resource as it is created from a .sadl file. However, the ...metrics.owl files are generated directly as OWL files so there is no .sadl file for them. By treating them as external resources they are added to the mapping file (ont-policy.rdf) and become usable--they can be queried, imported, etc.

**Comment by kittaakos** [comment]
The same rule should apply to any other `*.owl` resources generated from its `*.sadl` origin, right? If so, then we need to add such a filter to the external EMF resource factory.

**Comment by crapo** [comment]
Yes, I thought there was already one there, albeit one that sometimes is confused when a project is imported. Sven implemented a filter that I think is based on a file attribute. If a project is imported then OWL files that came from SADL files sometimes (I think) are treated as external.

**Comment by crapo** [comment]
You are right about the base URI on the Jena read call. I will refactor code and push to reflect what should be happening.

**Comment by kittaakos** [comment]
> You are right about the base URI on the Jena read call. I will refactor code and push to reflect what should be happening.

Please do not change the code if it is not urgent. I have a lot of outgoing changes which I have not committed yet. I would like to discuss the details first. Thanks for your understanding!

**Comment by crapo** [comment]
OK. The basic idea is that the downloadModels method in the UrlListEditor adds the mappings to the policy file. In the EmfExternalResource initModel one needs to get an absolute file URL from the resource URI, use the configMgr to get the publicURI, and that is the base URI for the call to the Jena method to read in the OntModel.

**Comment by kittaakos** [comment]
I have updated the [TODOs](https://github.com/crapo/sadlos2/issues/201#issue-234251312) on this task.

**Comment by kittaakos** [comment]
@crapo, in the external EMF tests, as discussed, I am creating one single shared `ConfigurationManagerForIDE` instance for the synthetic URIs and I add the mappings manually. This is the way I am trying to mimic the download functionality. For instance, the `altUrl` is `synthetic://test/Resource2.owl` and the public URI is `http://assert/Properties`, that happens before loading the external EMF resource. Unfortunately, the `com.ge.research.sadl.reasoner.ConfigurationManagerForEditing.getJenaDocumentMgr()` is `null`. What should I do to have the document manager initialized? Does it expect a `ont-policy.rdf` file on the disk?  Thank you!

**Comment by kittaakos** [comment]
@crapo, is there way to extract the base URI from the N3 or from the NT format without the configuration manager? Cannot we just use Jena to parse the content and get it? If so, we could get rid of the configuration manager from the external EMF resources.

**Comment by crapo** [comment]
Humm.. If you have added mappings then I would expect the act of adding a mapping would create a mapping model, and if there is a mapping model, then a call to getJenaDocumentMgr should create, if necessary, and return an OntDoucmentManager. So perhaps we will need to debug that tomorrow? 
As nearly as I can tell, N3 format can contain a base URI. N-triple format seems less certain. In either case, I don't think Jena provides access to the base URI. It seems that the base uri is at a lower, syntactic level than the semantic model. It is used to generate relative URIs, and so isn't seen as necessarily a part of the model. That's sort of consistent as seeing it as an identifier of the namespace. I think the n-triple format is more just complete URI's and you would have to organize concepts into namespaces by seeing which have the same base URI.
